corehq/apps/es/case_search.py

```diff
@@ -141,21 +141,25 @@ def case_property_filter(case_property_name, value):
     )
 
 
-def case_property_query(case_property_name, value, fuzzy=False):
+def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None):
     """
     Search for all cases where case property with name `case_property_name`` has text value `value`
     """
     if value is None:
         raise TypeError("You cannot pass 'None' as a case property value")
+    if multivalue_mode not in ['and', 'or', None]:
+        raise ValueError(" 'mode' must be one of: 'and', 'or', None")
     if value == '':
         return case_property_missing(case_property_name)
```

Comment:
Does multivalue_mode interact with this branch at all?  Ie, what should
`selected-all(my_prop, '')` and `selected-any(my_prop, '')` return?  As is, those will both match only cases where the user selected nothing.

Reply:
Could you clarify what you're asking a bit more for me? The value of `multivalue_mode` and the cases where a `selected` method is passed an empty string seem unrelated.

Reply:
In terms of how mobile handles this, it will return `True` if the case property contains a double space (or is empty):
```java
(" " + property_value + " ").contains(" " + input + " ")
```

It's not exactly obvious to me what the case search functionality should be.

Reply:
I think the options are to either:
1. reject queries with empty input
2. treat it as a 'missing property' query
3. do the actual multivalue query which will return no results

My intuition is that 1 is the most correct since 2 or 3 yield surprising results for a 'selected' query.

(and there isn't really a way to make this function the same as mobile without doing a regex query which I would say is out of scope).

Reply:
Waiting on feedback from AEs sounds good to me.  Definitely not a fan of implementing the mobile double-space behavior, since that doesn't sound deliberate.

> Could you clarify what you're asking a bit more for me? The value of `multivalue_mode` and the cases where a selected method is passed an empty string seem unrelated.

Simon elaborated a bit on what I meant, which is basically that there isn't an obvious correct behavior for what this function should do when using a multivalue mode and the value is the empty string.

For single select properties, it makes sense - if you say `favorite_color = 'purple'`, then it matches only those where the field is exactly that.  If you say `favorite_color = ''`, then you match only those where favorite_color is unspecified.

However, with multiselect it's more complicated.  If the caller says `selected-any("favorite_colors", "purple blue")`, then this should match cases where favorite_colors is "purple" or "red blue", but not "green yellow".  That much is straightforward.

But what if they say `selected-any("favorite_colors", "")`?  Does that match only users that haven't specified favorite_colors (the current behavior)?  All users?  None of them?  What about `selected-all`?

As an analogue, consider python.  This behavior seems intuitive to me, and may be a model we could follow here:
```python
In [1]: all(n in [1, 2, 3] for n in [])
Out[1]: True

In [2]: any(n in [1, 2, 3] for n in [])
Out[2]: False
```

Reply:
Thanks for the clarification, Ethan. I'm on the same page now.